explanation/intro-to/ebpf.md

```diff
@@ -0,0 +1,122 @@
+(introduction-to-ebpf)=
+# Introduction to eBPF
+
+eBPF is a lightweight virtual machine within the kernel. This virtual machine
+is exposed to the user-space and is able to safely run sandboxed programs in
+the kernel-space.
+
+BPF was originally designed to perform network packet filtering. Thus its name:
+_Berkley Packet Filter_. Over time, BPF was extended to use more registers,
+support 64-bit registers, data stores (Maps), and more. Such enhancement
+allowed this extended version of BPF (eBPF) to be decoupled from the kernel
+networking subsystem - eBPF was transformed in a tool with a broader scope,
+capable of enhancing not only the networking experience, but also tracing,
+profiling, observability, security, etc. Given its broader range of
+applications, calling eBPF a packet filter no longer describes its full
+capabilities. eBPF is now considered to be a standalone term and no longer
+refers to _extended Berkley Packet Filter_, reflecting its expanded role. Also,
+the terms eBPF and BPF are used interchangeably, to refer to eBPF.
+
+User-space applications can load eBPF programs into the kernel as eBPF
+bytecode. Although you could write eBPF programs in bytecode, there are several
+tools which provide abstraction layers on top of eBPF so you do not need to
+write bytecode manually. These tools will then generate the bytecode which will
+in turn be loaded into the kernel.
+
+Once the eBPF program is loaded, it is then verified by the kernel before it
+can run. These checks include verifying if the eBPF program halts and will
+never get stuck in a loop, verifying that the program will not crash by
+checking the registers and stack state validity throughout the program, and
+ensuring the process loading the eBPF program has all the capabilities required
+by the eBPF program to run.
+
+After the verification step, the bytecode is Just-In-Time (JIT) compiled into
+machine-code to optimize the program's execution.
+
+eBPF is a powerful tool for server and system administrators. It provides the
+ability to extract plenty of contextual information from systems or processes,
+offering several capabilities for monitoring, debugging, and securing systems
+with little overhead given it runs in kernel-space, i.e., there is no need for
+context-switching, making it very fast compared to solutions implemented in
+user-space.
+
```

Comment:
I'd add another subheading around here like "how it can be used"

Reply:
Done (different heading title though)